diff --git a/core/modules/node/src/NodeGrantDatabaseStorage.php b/core/modules/node/src/NodeGrantDatabaseStorage.php
index 155c480b8e..540e52358f 100644
--- a/core/modules/node/src/NodeGrantDatabaseStorage.php
+++ b/core/modules/node/src/NodeGrantDatabaseStorage.php
@@ -82,10 +82,17 @@ public function access(NodeInterface $node, $operation, AccountInterface $accoun
     $query->addExpression('1');
     // Only interested for granting in the current operation.
     $query->condition('grant_' . $operation, 1, '>=');
-    // Check for grants for this node and the correct langcode.
+    // Check for grants for this node and the correct langcode. New translations
+    // do not yet have a langcode and must check the fallback node record.
+    $langcode = $node->isNewTranslation() ? NULL : $node->language()->getId();
     $nids = $query->andConditionGroup()
-      ->condition('nid', $node->id())
-      ->condition('langcode', $node->language()->getId());
+      ->condition('nid', $node->id());
+    if (!is_null($langcode)) {
+      $nids->condition('langcode', $node->language()->getId());
+    }
+    else {
+      $nids->condition('fallback', 1);
+    }
     // If the node is published, also take the default grant into account. The
     // default is saved with a node ID of 0.
     $status = $node->isPublished();