diff --git a/core/includes/file.inc b/core/includes/file.inc
index c0be1e0253..29bbae45fa 100644
--- a/core/includes/file.inc
+++ b/core/includes/file.inc
@@ -65,6 +65,9 @@
  */
 function file_create_url($uri) {
   @trigger_error('file_create_url() is deprecated in drupal:9.3.0 and is removed from drupal:10.0.0. Use the appropriate method on \Drupal\Core\File\FileUrlGeneratorInterface instead. See https://www.drupal.org/node/2940031', E_USER_DEPRECATED);
+  if (is_null($uri)) {
+    return NULL;
+  }
   try {
     return \Drupal::service('file_url_generator')->generateAbsoluteString($uri);
   }
diff --git a/core/lib/Drupal/Core/Template/TwigExtension.php b/core/lib/Drupal/Core/Template/TwigExtension.php
index dfd20b264f..d7cc3beb10 100644
--- a/core/lib/Drupal/Core/Template/TwigExtension.php
+++ b/core/lib/Drupal/Core/Template/TwigExtension.php
@@ -107,7 +107,7 @@ public function getFunctions() {
       new TwigFunction('url', [$this, 'getUrl'], ['is_safe_callback' => [$this, 'isUrlGenerationSafe']]),
       new TwigFunction('path', [$this, 'getPath'], ['is_safe_callback' => [$this, 'isUrlGenerationSafe']]),
       new TwigFunction('link', [$this, 'getLink']),
-      new TwigFunction('file_url', [$this->fileUrlGenerator, 'generateString']),
+      new TwigFunction('file_url', [$this, 'getFileUrl']),
       new TwigFunction('attach_library', [$this, 'attachLibrary']),
       new TwigFunction('active_theme_path', [$this, 'getActiveThemePath']),
       new TwigFunction('active_theme', [$this, 'getActiveTheme']),
@@ -272,6 +272,22 @@ public function getLink($text, $url, $attributes = []) {
     return $build;
   }
 
+  /**
+   * Gets the file URL.
+   *
+   * @param string|null $uri
+   *   The file URI.
+   *
+   * @return string
+   *   The file URL.
+   */
+  public function getFileUrl(?string $uri) {
+    if (is_null($uri)) {
+      return NULL;
+    }
+    return $this->fileUrlGenerator->generateString($uri);
+  }
+
   /**
    * Gets the name of the active theme.
    *
diff --git a/core/modules/system/tests/modules/twig_theme_test/templates/twig_theme_test.file_url.html.twig b/core/modules/system/tests/modules/twig_theme_test/templates/twig_theme_test.file_url.html.twig
index 34ae66966b..aa5bb6f5ef 100644
--- a/core/modules/system/tests/modules/twig_theme_test/templates/twig_theme_test.file_url.html.twig
+++ b/core/modules/system/tests/modules/twig_theme_test/templates/twig_theme_test.file_url.html.twig
@@ -1 +1,2 @@
 <div>file_url: {{ file_url('core/modules/system/tests/modules/twig_theme_test/twig_theme_test.js') }}</div>
+<div>file_url: {{ file_url(undefined_variable) }}</div>
diff --git a/core/tests/Drupal/KernelTests/Core/File/FileSystemDeprecationTest.php b/core/tests/Drupal/KernelTests/Core/File/FileSystemDeprecationTest.php
index 6d3eb63c84..fb757a27a4 100644
--- a/core/tests/Drupal/KernelTests/Core/File/FileSystemDeprecationTest.php
+++ b/core/tests/Drupal/KernelTests/Core/File/FileSystemDeprecationTest.php
@@ -21,6 +21,7 @@ public function testDeprecatedFileCreateUrl() {
     $filepath = 'core/assets/vendor/jquery/jquery.min.js';
     $url = file_url_transform_relative(file_create_url($filepath));
     $this->assertNotEmpty($url);
+    $this->assertEquals(NULL, file_create_url(NULL));
   }
 
   /**
