diff --git a/core/modules/editor/src/Form/EditorImageDialog.php b/core/modules/editor/src/Form/EditorImageDialog.php
index 0e148d22f1..4fda54402e 100644
--- a/core/modules/editor/src/Form/EditorImageDialog.php
+++ b/core/modules/editor/src/Form/EditorImageDialog.php
@@ -12,6 +12,7 @@
 use Drupal\Core\Ajax\CloseModalDialogCommand;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Drupal\Core\Entity\EntityStorageInterface;
+use Drupal\Component\Render\PlainTextOutput;
 
 /**
  * Provides an image dialog for text editors.
@@ -96,10 +97,15 @@ public function buildForm(array $form, FormStateInterface $form_state, Editor $e
     $existing_file = isset($image_element['data-entity-uuid']) ? \Drupal::entityManager()->loadEntityByUuid('file', $image_element['data-entity-uuid']) : NULL;
     $fid = $existing_file ? $existing_file->id() : NULL;
 
+    $destination = trim($image_upload['directory'], '/');
+    // Replace tokens. As the tokens might contain HTML we convert it to plain
+    // text.
+    $destination = PlainTextOutput::renderFromHtml(\Drupal::token()->replace($destination));
+
     $form['fid'] = [
       '#title' => $this->t('Image'),
       '#type' => 'managed_file',
-      '#upload_location' => $image_upload['scheme'] . '://' . $image_upload['directory'],
+      '#upload_location' => $image_upload['scheme'] . '://' . $destination,
       '#default_value' => $fid ? [$fid] : NULL,
       '#upload_validators' => [
         'file_validate_extensions' => ['gif png jpg jpeg'],
