diff --git a/image_captcha/image_captcha.module b/image_captcha/image_captcha.module
index 7ad3c1a..90fa23f 100755
--- a/image_captcha/image_captcha.module
+++ b/image_captcha/image_captcha.module
@@ -92,7 +92,7 @@ function _image_captcha_get_available_fonts_from_directories($directories = NULL
   // Collect the font information.
   $fonts = [];
   foreach ($directories as $directory) {
-    if (\Drupal::service('file_system')->prepareDirectory($directory)) {
+    if (is_dir($directory) && is_readable($directory)) {
       $files = \Drupal::service('file_system')->scanDirectory($directory, '/\.[tT][tT][fF]$/');
       foreach ($files as $filename => $font) {
         $fonts[hash('sha256', $filename)] = (array) $font;
