diff --git a/core/modules/path/src/Plugin/Validation/Constraint/PathAliasConstraintValidator.php b/core/modules/path/src/Plugin/Validation/Constraint/PathAliasConstraintValidator.php
index be6350f75..c079a9b0e
--- a/core/modules/path/src/Plugin/Validation/Constraint/PathAliasConstraintValidator.php
+++ b/core/modules/path/src/Plugin/Validation/Constraint/PathAliasConstraintValidator.php
@@ -49,6 +49,9 @@ public function validate($value, Constraint $constraint) {
       /** @var \Drupal\Core\Entity\ContentEntityInterface $original */
       $original = $this->entityTypeManager->getStorage($entity->getEntityTypeId())->loadUnchanged($entity->id());
       $entity_langcode = $entity->language()->getId();
+      if ($original->language()->getId() !== $entity_langcode && $original->hasTranslation($entity_langcode)) {
+        $original = $original->getTranslation($entity_langcode);
+      }
 
       // Only add the violation if the current translation does not have the
       // same path alias.
