diff --git a/core/lib/Drupal/Core/Updater/Module.php b/core/lib/Drupal/Core/Updater/Module.php
index c7cbb7f..68d705d 100644
--- a/core/lib/Drupal/Core/Updater/Module.php
+++ b/core/lib/Drupal/Core/Updater/Module.php
@@ -40,7 +40,20 @@ public function getInstallDirectory() {
    * {@inheritdoc}
    */
   public static function getRootDirectoryRelativePath() {
-    return 'modules';
+      // are we in multi-site?
+      if (file_exists(DRUPAL_ROOT . '/sites/sites.php')) {
+        $my_dir = config_get_config_directory('sync');
+        $my_regex = "/(sites\/[0-9a-zA-Z-\.]+)\/.*/";
+        $my_matches = array();
+        preg_match($my_regex, $my_dir, $my_matches);
+        $my_match='';
+        if (count($my_matches) > 1) {
+          $my_match = $my_matches[1];
+        }
+        return "${my_match}/modules";
+      } else {
+        return "modules";
+      }
   }
 
   /**
diff --git a/core/lib/Drupal/Core/Updater/Theme.php b/core/lib/Drupal/Core/Updater/Theme.php
index 5dfbd97..30b12d6 100644
--- a/core/lib/Drupal/Core/Updater/Theme.php
+++ b/core/lib/Drupal/Core/Updater/Theme.php
@@ -40,7 +40,20 @@ public function getInstallDirectory() {
    * {@inheritdoc}
    */
   public static function getRootDirectoryRelativePath() {
-    return 'themes';
+    // are we in multi-site?
+    if (file_exists(DRUPAL_ROOT . '/sites/sites.php')) {
+      $my_dir = config_get_config_directory('sync');
+      $my_regex = "/(sites\/[-_0-9a-zA-Z-\.]+)\/.*/";
+      $my_matches = array();
+      preg_match($my_regex, $my_dir, $my_matches);
+      $my_match='';
+      if (count($my_matches) > 1) {
+        $my_match = $my_matches[1];
+      }
+      return "${my_match}/themes";
+    } else {
+        return "themes";
+    }
   }
 
   /**
