diff --git a/core/lib/Drupal/Core/Updater/Module.php b/core/lib/Drupal/Core/Updater/Module.php
index 4829c53..d36d0c8 100644
--- a/core/lib/Drupal/Core/Updater/Module.php
+++ b/core/lib/Drupal/Core/Updater/Module.php
@@ -40,7 +40,7 @@ public function getInstallDirectory() {
    * {@inheritdoc}
    */
   public static function getRootDirectoryRelativePath() {
-    return 'modules';
+    return '/sites/default/modules';
   }
 
   /**
diff --git a/core/lib/Drupal/Core/Updater/Theme.php b/core/lib/Drupal/Core/Updater/Theme.php
index 4b6cd40..7ac2495 100644
--- a/core/lib/Drupal/Core/Updater/Theme.php
+++ b/core/lib/Drupal/Core/Updater/Theme.php
@@ -40,7 +40,7 @@ public function getInstallDirectory() {
    * {@inheritdoc}
    */
   public static function getRootDirectoryRelativePath() {
-    return 'themes';
+    return '/sites/default/themes';
   }
 
   /**
