diff --git a/src/Plugin/CKEditorPlugin/PanelButton.php b/src/Plugin/CKEditorPlugin/PanelButton.php
index c9c239f..48770bd 100644
--- a/src/Plugin/CKEditorPlugin/PanelButton.php
+++ b/src/Plugin/CKEditorPlugin/PanelButton.php
@@ -19,7 +19,7 @@ class PanelButton extends CKEditorPluginBase {
    * {@inheritdoc}
    */
   public function getFile() {
-    $path = '/libraries/panelbutton/plugin.js';
+    $path = 'libraries/panelbutton/plugin.js';
     if (\Drupal::moduleHandler()->moduleExists('libraries')) {
       $path = libraries_get_path('panelbutton') . '/plugin.js';
     }

