diff --git a/core/modules/views/src/Plugin/views/display/PathPluginBase.php b/core/modules/views/src/Plugin/views/display/PathPluginBase.php
index ab242ff773..445150053d 100644
--- a/core/modules/views/src/Plugin/views/display/PathPluginBase.php
+++ b/core/modules/views/src/Plugin/views/display/PathPluginBase.php
@@ -507,6 +507,10 @@ protected function validatePath($path) {
       $errors[] = $this->t('Invalid path. Valid characters are alphanumerics as well as "-", ".", "_" and "~".');
     }
 
+    if (UrlHelper::isExternal($path)) {
+      $errors[] = $this->t("Path component '{$path}' is external. External URL is not accepted as view path.");
+    }
+
     $path_sections = explode('/', $path);
     // Symfony routing does not allow to use numeric placeholders.
     // @see \Symfony\Component\Routing\RouteCompiler
diff --git a/core/modules/views/tests/src/Functional/Plugin/DisplayTest.php b/core/modules/views/tests/src/Functional/Plugin/DisplayTest.php
index eea496beea..5bcb407881 100644
--- a/core/modules/views/tests/src/Functional/Plugin/DisplayTest.php
+++ b/core/modules/views/tests/src/Functional/Plugin/DisplayTest.php
@@ -458,4 +458,27 @@ protected function checkTranslationSetting($expected_node_translatability = FALS
     $this->assertNoText($supported_text);
   }
 
+  /**
+   * Tests view with external page url should throw error.
+   */
+  public function testViewUrlAsExternal() {
+    $this->drupalGet('admin/structure/views/add');
+    $this->assertSession()->statusCodeEquals(200);
+    // Create a view that sorts oldest first.
+    $view1 = [];
+    $view1['label'] = $this->randomMachineName(16);
+    $view1['id'] = strtolower($this->randomMachineName(16));
+    $view1['description'] = $this->randomMachineName(16);
+    $view1['page[create]'] = 1;
+    $view1['page[title]'] = $this->randomMachineName(16);
+    $view1['page[path]'] = 'https://www.' . $this->randomMachineName(16) . '.com';
+    $this->drupalPostForm('admin/structure/views/add', $view1, 'Save and edit');
+    $this->assertRaw('Path component ' . "'" . $view1['page[path]'] . "'" . ' is external. External URL is not accepted as view path.');
+
+    // Saving view with internal url.
+    $view1['page[path]'] = 'admin/' . $this->randomMachineName(16);
+    $this->drupalPostForm('admin/structure/views/add', $view1, 'Save and edit');
+    $this->assertRaw(t('The view %view has been saved.', ['%view' => $view1['label']]));
+  }
+
 }
